File: cpp/src/HexStringEncoder.cpp


#include "HexStringEncoder.h"

std::string HexStringEncoder::encode(ByteBuffer& buffer) {
    std::string str;
    while (buffer.getAvailable() > 0) {
        int octet = buffer.read();
        int tmp1 = (octet >> 4) & 0x0F;
        int tmp2 = octet & 0x0F;
        char    char1;
        char    char2;
        if (tmp1 <= 9)
        {
            char1 = tmp1 + '0';
        }
        else
        {
            char1 = tmp1 - 10 + 'a';
        }
        if (tmp2 <= 9)
        {
            char2 = tmp2 + '0';
        }
        else
        {
            char2 = tmp2 - 10 + 'a';
        }
        std::string str3 = "00";
        str3[0] = char1;
        str3[1] = char2;
        str.append(str3);
    }

    return str;
}

File: swift/Sources/Encode-TestDesign-Kata/HexStringEncoder.swift
class HexStringEncoder {

    func encode(_ buffer: ByteBuffer) -> String {
        var str = ""
        while buffer.getAvailable() > 0 {
            if let octet = buffer.read() {
                let tmp1 = (octet >> 4) & 0x0F
                let tmp2 = octet & 0x0F

                let char1: Character
                let char2: Character

                if tmp1 <= 9 {
                    char1 = Character(UnicodeScalar(tmp1 + 48)!) // ASCII for '0'
                } else {
                    char1 = Character(UnicodeScalar(tmp1 - 10 + 97)!) // ASCII for 'a'
                }

                if tmp2 <= 9 {
                    char2 = Character(UnicodeScalar(tmp2 + 48)!) // ASCII for '0'
                } else {
                    char2 = Character(UnicodeScalar(tmp2 - 10 + 97)!) // ASCII for 'a'
                }

                str.append(char1)
                str.append(char2)
            }
        }

        return str
    }
}
